Allow Recursive Schema Creation

Ademan opened the pull request:
This is #25 rebased against master with a few minor changes to get things working, a couple of spelling corrections, and the extraneous tests/bak file not included from the original PR. I couldn't see a way to attach this directly to the original, so I created this new pull request. All tests are passing currently.

Ademan commented:
I found pretty quickly there is an untested case which causes an error in my existing code, I will prepare an appropriate additional test and patch to add to this PR.

Ademan commented:
The test in 1a8d2f6 is as simple as I could make it, it seems the issue requires traversing a relationship before traversing a backref. The event listener changes in c54b3bb were necessary to prevent a race condition being introduced in 1a8d2f6 which occasionally causes all tests to fail, but still allows testing of the mapper_configured event. These changes are ready for review.

tisdall commented:
Thanks for working on this.  It's been a long standing issue.  I don't really have any time to take a look at this in detail, though.

tisdall commented on colanderalchemy/schema.py:
```diff
@@ -113,6 +114,11 @@ class MyModel(Base):
+           An ``SQLAlchemySchemaNode`` parent list to avoid relationship circular
+           dependencies and thus prevent infinite recursion. Used internally.
+
+           Default: []
```
Should we be documenting an argument that shouldn't really be used externally?  Is there any use for this externally?

Ademan replied:
> Should we be documenting an argument that shouldn't really be used externally?

You're right, probably not.

> Is there any use for this externally?

I actually am already *abusing* this to filter certain relationships. I pass in my `User` model to the `parents_` argument to prevent ColanderAlchemy from adding any relationships to `User` in the schema. I have ideas and code for an alternative way to filter columns and relationships, so the use case I described probably shouldn't be supported.

tisdall commented:
Would someone using relationships like to give this a look-over before I merge it?

davidjb commented:
@tisdall Happy to test on my code.  @Ademan can you resolve the merge conflicts and rebase on the latest master?

Ademan commented:
There, rebased. Tests still passing locally, Travis CI should agree. Assuming that passes, these changes are ready for review again. @tisdall I have no strong opinion regarding the parents_ documentation, do you want me to remove it?

tisdall commented:
@Ademan I just thought of one last issue...  If you have `parents_` like that, I think it'll show up in the API documentation.  If we really want to hide it, we need to make it part of the `**kw` element.

I'd prefer to hide the `parents_` element completely from the end-user.  So, can you remove the documentation and also move the argument into the `**kw` element?  I think that would make it `self.parents_ = kwargs.pop('parents_', [])` but placed before the call to `super`.

davidjb commented:
Tested and working fine with my existing relationship codebase.

tisdall commented:
sweet!  One last thing...  Add an entry to CHANGES.rst and I'll merge it.

Ademan commented:
Rebased again, everything still looks fine, added CHANGES.rst.

Two brief comments

1. Previously, any child `SQLAlchemySchemaNode`'s `class_` attribute was a SQLAlchemy Mapper, while the root `class_` was always (in my opinion, correctly) the mapped class itself. This change set changes this, any code that pokes at `class_` of child nodes may be affected. In my opinion the old behavior was incorrect (and was inconsistent), and `class_` should not be accessed by users, but this is a risk worth mentioning.
2. These changes change the *default* behavior of `SQLAlchemySchemaNode`s to recurse into child models. Should we disable this behavior by default?

tisdall commented:
I think I'm using `class_` in my code..  ^_^   I'm unclear on what you're saying, though.  The root object has `class_` being the mapped class but the children were the mapper and now you've changed it so it's also the mapped class?  Is that right?

As for 2, wasn't it already going into relationship classes and adding those by default?  These changes just make it so it extends down the chain as far as it needs to instead of stopping at a particular depth.

Ademan commented:
> children were the mapper and now you've changed it so it's also the mapped class? Is that right?

Correct.

> As for 2 [...] These changes just make it so it extends down the chain as far as it needs to instead of stopping at a particular depth.

Also correct, I just wanted to be sure we're ok with that, I could imagine someone upgrading and suddenly directly exposing their user model or something like that.

tisdall commented:
Well, maybe these changes are enough to make the next version 0.4
